JDK-8266536: Provide a variant of os::iso8601_time which works with arbitrary timestamps

tstuefe opened the pull request:
We have os::iso8601_time(), which gives an ISO8601 timestamp of the current time. It would be very useful to have a second variant which can be fed an arbitrary numerical timestamp.

This is useful in the context of making asynchronous UL logging cheaper (see JDK-8229517)

This patch provides an additional API: 
`char* os::iso8601_time(jlong milliseconds_since_19700101, char* buffer, size_t buffer_length, bool utc);`
alongside the existing
`char* os::iso8601_time(char* buffer, size_t buffer_length, bool utc);`
and implements the latter using the former. Not much code added.

In addition, it adds a regression gtest for these APIs.

Please ignore the harfbuzz change, its a build fix needed for older gcc, will be removed before final push.

Testing: GHA, manual gtests, SAP nightlies on all our platforms.

Thanks, Thomas

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8266536](https://bugs.openjdk.java.net/browse/JDK-8266536): Provide a variant of os::iso8601_time which works with arbitrary timestamps


### Reviewers
 * [Xin Liu](https://openjdk.java.net/census#xliu) (@navyxliu - no project role) ⚠️ Review applies to bfa41f13d89bbfee3726c8468be36e9c62e21dbc
 * [Volker Simonis](https://openjdk.java.net/census#simonis) (@simonis - **Reviewer**) ⚠️ Review applies to 40f23a2324a61f4e407440ea6655cdfff18f6edb
 * [Yasumasa Suenaga](https://openjdk.java.net/census#ysuenaga) (@YaSuenag - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3869/head:pull/3869` \
`$ git checkout pull/3869`

Update a local copy of the PR: \
`$ git checkout pull/3869` \
`$ git pull https://git.openjdk.java.net/jdk pull/3869/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3869`

View PR using the GUI difftool: \
`$ git pr show -t 3869`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3869.diff">https://git.openjdk.java.net/jdk/pull/3869.diff</a>

</details>

navyxliu commented on src/java.desktop/share/native/libharfbuzz/hb-ot-shape-complex-use-machine.hh:
```diff
@@ -375,7 +375,7 @@ hb_iter_with_fallback_t<machine_index_t<Iter>,
 {
         machine_index_t (const Iter& it) : it (it) {}
-        machine_index_t (const machine_index_t& o) : it (o.it) {}
+        machine_index_t (const machine_index_t& o) : hb_iter_with_fallback_t<machine_index_t<Iter>, typename Iter::item_t>(o), it (o.it) {}
```
is it relevant?  
I feel it belongs to a separated issue.

tstuefe replied:
Sorry, forgot to mention. Unrelated harfbuzz build fix for gcc 8 and earlier. Will remove before the final push.

tstuefe replied:
> The API change looks good to me. I understand that this is prerequisite of your logdecoration change. make sense.
> 
> I don't understand your change in harfbuzz header.

Thanks for your review, Xin.

navyxliu reviewed (approved):
The API change looks good to me. I understand that this is prerequisite of your logdecoration change. make sense.

I don't understand your change in harfbuzz header.

simonis commented on test/hotspot/gtest/runtime/test_os.cpp:
```diff
@@ -754,3 +754,62 @@ TEST_VM(os, dll_address_to_function_and_library_name) {
+static bool very_simple_string_matcher(const char* pattern, const char* s) {
+  const size_t lp = strlen(pattern);
+  const size_t ls = strlen(s);
+  if (ls < lp) {
```
Shouldn't this be `!=`?

tstuefe replied:
I wanted to be able to give prefix pattern which do not cover the whole string.

simonis commented on test/hotspot/gtest/runtime/test_os.cpp:
```diff
@@ -754,3 +754,62 @@ TEST_VM(os, dll_address_to_function_and_library_name) {
+  EXPECT_EQ(result, buffer);
+  EXPECT_TRUE(very_simple_string_matcher(pattern, result));
+
+  result = os::iso8601_time(buffer, sizeof(buffer), false);
+  tty->print_cr("%s", result);
+  EXPECT_EQ(result, buffer);
+  EXPECT_TRUE(very_simple_string_matcher(pattern, result));
```
Do you intentionally repeat this same test for a second time? Does it provide any benefit?

tstuefe replied:
Oops, no, this is an oversight. I'll remove the second test, thanks for catching this.

simonis reviewed (changes requested):
Change looks good. I have just two minor questions (see inline).

Thanks for adding the gtest!

tstuefe commented:
> Change looks good. I have just two minor questions (see inline).
> 
> Thanks for adding the gtest!

Thanks Volker!

simonis reviewed (approved):
Looks good, thanks.

tstuefe commented:
> Looks good, thanks.

Thank you Volker!

navyxliu commented on src/hotspot/share/runtime/os.hpp:
```diff
@@ -224,6 +224,13 @@ class os: AllStatic {
+  // Fill in buffer with an ISO-8601 string corresponding to the given javaTimeMillis value
+  // E.g., YYYY-MM-DDThh:mm:ss.mmm+zzzz.
+  // Returns buffer, or NULL if it failed.
+  static char* iso8601_time(jlong milliseconds_since_19700101, char buffer[29],
```
You are supposed to get a compiler warning on this char buffer[29]. ISO C/C++ can't pass in array type as a function argument.  As a result, the type buffer automatically casts to char* here. 

Further, the function signature is different from its implementation, which use char* buffer instead. 
I see that you do enforce length checking in implementation.

tstuefe replied:
Oh, good catch, this is a stray edit I thought I reverted. See below.

> ISO C/C++ can't pass in array type as a function argument. As a result, the type buffer automatically casts to char* here.

No, it can, that's totally fine. It gets automatically translated to pointer type, but its a valid expression nonetheless. 

It is usually done to indicate that a function expects an output array of a particular length instead of letting the caller guess the right length - and letting him to deal with truncation if he guesses wrong.

I originally planned to define the function as 
```
iso8601_time(jlong milliseconds_since_19700101, char buffer[29]);
```
to indicate that the only output value length which makes sense is 29, but then I refrained because I wanted to keep the patch simple.

navyxliu replied:
C/C++  allow you to write code like that, but they won't check the length of argument <=29 for you. 
I think you sanity check inside of iso8601_time() is good enough.

tstuefe commented:
Thanks @navyxliu and @simonis and @YaSuenag for your reviews.

Last change:
- fixes the stray edit in the prototype Xin noted
- adds an explicit constant for the output buffer size of iso timestamps

@YaSuenag could you please officially approve this PR? Unfortunately, Xin is no committer yet, so I need a second reviewer.

Thanks, Thomas

YaSuenag reviewed (approved):
Looks good!
I think `needed_buffer` in `os::iso8601_time()` is no longer needed, but it's ok if it is left.

tstuefe commented:
> Looks good!
> I think `needed_buffer` in `os::iso8601_time()` is no longer needed, but it's ok if it is left.

Thanks Yasumasa! I changed the code a bit and removed needed_buffer.

YaSuenag reviewed (approved):
Thank you for fixing it! It still looks good.
